lib/src/memory/cam.dart

```diff
@@ -0,0 +1,142 @@
+
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// An interface to a Content-Addressable Memory (CAM) that allows querying
+/// for tags and returns the index of the matching tag if found.
+class TagInterface extends Interface<DataPortGroup> {
+  /// The width of data in the memory.
+  final int tagWidth;
+
+  /// The width of addresses in the memory.
+  final int idWidth;
+
+  /// The "enable" bit for this interface, enabling a request.
+  Logic get en => port('en');
+
+  /// The "tag" to match when [en] is high.
+  Logic get tag => port('tag');
+
+  /// The entry number (index) where the tag was found.
+  Logic get idx => port('idx');
+
+  /// If the tag was found a 'hit' is indicated.
+  Logic get hit => port('hit');
+
+  /// Constructs a new interface of specified [idWidth] and [tagWidth] for
+  /// querying a CAM.
+  TagInterface(this.idWidth, this.tagWidth) {
+    setPorts([
+      Logic.port('en'),
+      Logic.port('tag', tagWidth),
+    ], [
+      DataPortGroup.control
+    ]);
+
+    setPorts([
+      Logic.port('idx', idWidth),
+      Logic.port('hit'),
+    ], [
+      DataPortGroup.data
+    ]);
+  }
+
+  /// Makes a copy of this [TagInterface] with matching configuration.
+  @override
+  TagInterface clone() => TagInterface(idWidth, tagWidth);
+}
+
+/// A content-addressable memory (CAM) module.
+class Cam extends Memory {
+  /// The number of entries in the Cam.
+  final int numEntries;
+
+  /// The number of lookup (tag access) ports.
+  final int numLookups;
+
+  /// The lookup (tag access) ports.
+  @protected
+  final List<TagInterface> lookupPorts = [];
+
+  /// Constructs a new [Cam] with write ports that use direct address writes
+  /// and .read ports that use associative lookup.
+  Cam(Logic clk, Logic reset, List<DataPortInterface> writePorts,
+      List<TagInterface> lookupPorts,
+      {this.numEntries = 8,
+      super.name = 'cam',
+      super.reserveName,
+      super.reserveDefinitionName,
+      String? definitionName})
+      : numLookups = lookupPorts.length,
+        super(clk, reset, writePorts, [],
+            definitionName: definitionName ??
+                'CAM_WP${writePorts.length}'
+                    '_LP${lookupPorts.length}_E$numEntries') {
+    if (lookupPorts.isEmpty) {
+      throw ArgumentError('Must provide at least one lookup port.');
+    }
+    for (var i = 0; i < numLookups; i++) {
+      this.lookupPorts.add(lookupPorts[i].clone()
+        ..connectIO(this, lookupPorts[i],
+            inputTags: {DataPortGroup.control},
+            outputTags: {DataPortGroup.data},
+            uniquify: (original) => 'lookup_${original}_$i'));
+    }
+    _buildLogic();
+  }
+
+  /// Flop-based storage of all memory.
+  late final List<Logic> _storageBank;
+
+  void _buildLogic() {
+    // create local storage bank
+    _storageBank = List<Logic>.generate(
```

Comment:
we could make this a `LogicArray` instead (planned on fixing the RF also https://github.com/intel/rohd-hcl/issues/256, https://github.com/intel/rohd-hcl/pull/258)